arelle/plugin/validate/HMRC/ValidateHmrc.py

```diff
@@ -331,13 +347,54 @@ def _lang(self) -> HmrcLang:
                         return HmrcLang.WELSH
         return HmrcLang.ENGLISH
 
+    def _verifyAuditFacts(self, prefix: str) -> None:
+        """
+        Logs an error when an audited report does not facts tagged with the concepts of "DateAuditorsReport",  "OpinionAuditorsOnEntity"
+        as well as either "NameIndividualAuditor" or the combination of "NameSeniorStatutoryAuditor" and "NameEntityAuditors".
+        :param prefix:
+        :return:
+        """
+        missingConcepts = []
+        dateFacts = self._getFacts(CONCEPT_DATE_AUDITOR_REPORT)
+        if not any(self._checkValidFact(x) for x in dateFacts):
+            missingConcepts.append(CONCEPT_DATE_AUDITOR_REPORT)
+        opinionFacts = self._getFacts(CONCEPT_OPINION_AUDITORS_ON_ENTITY)
+        if not any(self._checkValidFact(x) for x in opinionFacts):
+            missingConcepts.append(CONCEPT_OPINION_AUDITORS_ON_ENTITY)
+        indFacts = self._getFacts(CONCEPT_NAME_INDIVIDUAL_AUDITOR)
+        senFacts = self._getFacts(CONCEPT_NAME_SENIOR_STATUTORY_AUDITOR)
+        entFacts = self._getFacts(CONCEPT_NAME_ENTITY_AUDITORS)
+        if not (any(self._checkValidFact(x) for x in entFacts) and any(self._checkValidFact(x) for x in senFacts)) and not any(self._checkValidFact(x) for x in indFacts):
+            missingConcepts.append(CONCEPT_NAME_INDIVIDUAL_AUDITOR)
+            missingConcepts.append(CONCEPT_NAME_SENIOR_STATUTORY_AUDITOR)
+            missingConcepts.append(CONCEPT_NAME_ENTITY_AUDITORS)
+        if len(missingConcepts) > 0:
+            self.modelXbrl.error(
+                prefix + ".Audit",
+                "An audited report must contain facts tagged with the concepts of DateAuditorsReport, OpinionAuditorsOnEntity"
+                "as well as either NameIndividualAuditor or the combination of NameSeniorStatutoryAuditor and NameEntityAuditors"
+                "There are no facts tagged with the concepts: %(concepts)s",
+                concepts=missingConcepts,
+            )
+
+    def _verifyMedFact(self, circumfix: str) -> None:
```

Comment:
In line with my `_verifyAuditFacts` comment above, this logic could be added to `_evaluateCode`. The logic seems pretty generic (evaluating given code should check if given fact exists and is non-nil) so would probably eventually be duplicated for similar errors.

You could establish some kind of `CODE_REQUIRED_CONCEPTS` dict that maps codes to concept local names (i.e. `CODE_REQUIRED_CONCEPTS = { CO_MED_CO: (CONCEPT_MEDIUM_COMPANY,) }`).
Then in the validation you'd have 
```
result = self._evaluateCode(CO_MED_CO)
if not result.success:
  self._errorOnMissingFact(CO_MED_CO, result)
```